File: ts/tests/conformance.test.ts
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import yaml from "js-yaml";
import { describe, expect, test } from "vitest";
import { ExecAdapter, type ExecRequest } from "../src/adapters/exec.js";
import { FsAdapter, type FsRequest } from "../src/adapters/fs.js";
import { HttpAdapter, type HttpRequest } from "../src/adapters/http.js";
import { RedisAdapter, type RedisRequest } from "../src/adapters/redis.js";
import { SqlAdapter, type SqlRequest } from "../src/adapters/sql.js";
import { FileCassette } from "../src/cassette.js";
import { FileSession } from "../src/session.js";
import { StreamFormatError, type StreamedInteraction } from "../src/stream.js";
import { counterStreamFingerprint, serverStreamFingerprint } from "../src/streamfp.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const fixturesRoot = path.resolve(__dirname, "../../spec/fixtures");

interface ManifestInteraction {
  adapter: string;
  fingerprint: string;
  streamed?: boolean;
  /**
   * Marks a unary entry whose fingerprint is a computed value: the walker
   * rebuilds the adapter's request from the req payload and recomputes it
   * with the adapter's algorithm.
   */
  verify_fingerprint?: boolean;
}

interface Manifest {
  interactions: ManifestInteraction[];
}

interface GrpcStreamPayload {
  service: string;
  method: string;
  n?: number;
}

function tmpDir(): string {
  return fs.mkdtempSync(fs.realpathSync("/tmp") + "/xrr-conformance-");
}

/**
 * Round-trip a streamed pair: re-emit into a fresh dir, load back, compare
 * field-for-field. Frame messages are held as decoded bytes in the model, so
 * deep equality compares decoded bytes; YAML formatting and encoding choice
 * are free per spec.
 */
async function roundTrip(pair: StreamedInteraction): Promise<void> {
  const out = new FileCassette(tmpDir());
  await out.saveStreamed(pair);
  const reloaded = await out.loadStreamed(pair.req.adapter, pair.req.fingerprint);
  expect(reloaded).toEqual(pair);
}

/**
 * Recompute a grpc streaming fingerprint from the loaded pair per spec.
 *
 * `n` comes from each pair's own payload rather than from a counter shared
 * across the dir, so the manifest loop below is order-independent: the spec's
 * ordering rule (cassette-format-streaming.md, Manifest Extension) holds
 * vacuously here, with no shared counter for a wrong order to desynchronise.
 * The "manifest order is irrelevant" test pins that so the construction cannot
 * regress into a shared-counter loop unnoticed.
 */
function recomputeGrpcFingerprint(pair: StreamedInteraction): string {
  const payload = pair.req.payload as GrpcStreamPayload;
  if (pair.req.stream.type === "server") {
    return serverStreamFingerprint(payload.service, payload.method, pair.req.stream.frames[0].message);
  }
  // client/bidi: the informational payload `n` documents the occurrence.
  return counterStreamFingerprint(pair.req.stream.type, payload.service, payload.method, payload.n ?? 0);
}

/**
 * Recompute a unary fingerprint from the loaded req payload with the
 * adapter's own algorithm. Loading alone cannot expose a canonical-JSON
 * escaping fork — the files load in every port; the derived key is what
 * differs.
 */
async function recomputeUnaryFingerprint(adapter: string, payload: unknown): Promise<string> {
  switch (adapter) {
    case "exec":
      return new ExecAdapter().fingerprint(payload as ExecRequest);
    case "http":
      return new HttpAdapter().fingerprint(payload as HttpRequest);
    case "sql":
      return new SqlAdapter().fingerprint(payload as SqlRequest);
    case "fs":
      return new FsAdapter().fingerprint(payload as FsRequest);
    case "redis":
      return new RedisAdapter().fingerprint(payload as RedisRequest);
    default:
      throw new Error(`verify_fingerprint: no unary fingerprint model for adapter ${adapter}`);
  }
}

/** Loads one unary pair (a miss throws) and verifies a pinned fingerprint. */
async function conformUnaryPair(
  cassette: FileCassette,
  interaction: ManifestInteraction
): Promise<void> {
  const { req } = await cassette.load(interaction.adapter, interaction.fingerprint);
  if (interaction.verify_fingerprint) {
    expect(await recomputeUnaryFingerprint(interaction.adapter, req)).toBe(interaction.fingerprint);
  }
}

describe("conformance fixtures", () => {
  const entries = fs.readdirSync(fixturesRoot, { withFileTypes: true });
  const dirs = entries.filter((e) => e.isDirectory());

  expect(dirs.length).toBeGreaterThan(0);

  for (const entry of dirs) {
    const fixtureDir = path.join(fixturesRoot, entry.name);
    const manifestPath = path.join(fixtureDir, "manifest.yaml");

    test(entry.name, async () => {
      const raw = fs.readFileSync(manifestPath, "utf8");
      const manifest = yaml.load(raw) as Manifest;
      const cassette = new FileCassette(fixtureDir);

      for (const interaction of manifest.interactions) {
        if (interaction.streamed) {
          const pair = await cassette.loadStreamed(interaction.adapter, interaction.fingerprint);
          expect(pair.req.fingerprint).toBe(interaction.fingerprint);
          expect(pair.resp.fingerprint).toBe(interaction.fingerprint);
          await roundTrip(pair);
          if (interaction.adapter === "grpc") {
            expect(recomputeGrpcFingerprint(pair)).toBe(interaction.fingerprint);
          }
        } else {
          await conformUnaryPair(cassette, interaction);
        }
      }
    });
  }
});

describe("manifest order is irrelevant", () => {
  // `interactions` is an unordered set (Manifest Extension), so reversing a
  // manifest's entries is a legal edit and must not change any result. Only
  // grpc-client-stream-repeat has entries sharing a counter domain, so it is
  // the dir a shared-counter regression would break first.
  const dirs = fs
    .readdirSync(fixturesRoot, { withFileTypes: true })
    .filter((e) => e.isDirectory());

  for (const entry of dirs) {
    test(entry.name, async () => {
      const fixtureDir = path.join(fixturesRoot, entry.name);
      const manifest = yaml.load(
        fs.readFileSync(path.join(fixtureDir, "manifest.yaml"), "utf8")
      ) as Manifest;
      const cassette = new FileCassette(fixtureDir);

      for (const interaction of [...manifest.interactions].reverse()) {
        if (interaction.streamed) {
          const pair = await cassette.loadStreamed(interaction.adapter, interaction.fingerprint);
          expect(pair.req.fingerprint).toBe(interaction.fingerprint);
          if (interaction.adapter === "grpc") {
            expect(recomputeGrpcFingerprint(pair)).toBe(interaction.fingerprint);
          }
        } else {
          await conformUnaryPair(cassette, interaction);
        }
      }
    });
  }
});

describe("streamed conformance details", () => {
  test("grpc-stream-malformed-b64 pair is rejected, not decoded leniently", async () => {
    // Deliberately absent from its manifest: harnesses target the pair by
    // path and assert strict loading fails (fixture README).
    const cassette = new FileCassette(path.join(fixturesRoot, "grpc-stream-malformed-b64"));
    await expect(cassette.loadStreamed("grpc", "8dbfb222")).rejects.toThrow(StreamFormatError);
    await expect(cassette.loadStreamed("grpc", "8dbfb222")).rejects.toThrow(/base64/);
  });

  test("sse-text-scalars hazard payloads decode to exact characters", async () => {
    const cassette = new FileCassette(path.join(fixturesRoot, "sse-text-scalars"));
    const pair = await cassette.loadStreamed("sse", "66ecc77a");
    const texts = pair.resp.stream.frames.map((f) => new TextDecoder().decode(f.message));
    expect(texts).toEqual(["on", "12:30", "null", " leading", "trailing ", "  padded  "]);
  });

  test("grpc-stream-error keeps the v1 error field and status payload", async () => {
    const cassette = new FileCassette(path.join(fixturesRoot, "grpc-stream-error"));
    const pair = await cassette.loadStreamed("grpc", "9e8c4d4c");
    expect(pair.resp.error).toBe("rpc error: code = Unavailable desc = connection reset");
    expect(pair.resp.payload).toEqual({ status_code: 14 });
    expect(pair.resp.stream.frames).toHaveLength(2);
    expect(pair.resp.stream.end.seq).toBe(4);
  });

  test("grpc-client-stream-repeat: scripted two-open session yields n=0 then n=1", async () => {
    // The spec's n=1 obligation: a second open of the same tuple within one
    // session. The session's occurrence counter drives the fingerprints; the
    // fixture dir supplies the static cassette material.
    const dir = path.join(fixturesRoot, "grpc-client-stream-repeat");
    const cassette = new FileCassette(dir);
    const session = new FileSession("replay", cassette);

    const open = async (service: string, method: string) => {
      const n = session.streamCounter.next(service, method, "client");
      const fp = counterStreamFingerprint("client", service, method, n);
      const pair = await cassette.loadStreamed("grpc", fp);
      return { n, fp, pair };
    };

    const first = await open("files.FileService", "Upload");
    expect(first.n).toBe(0);
    expect(first.fp).toBe("2bebfd6f");
    expect((first.pair.req.payload as GrpcStreamPayload).n).toBe(0);

    const second = await open("files.FileService", "Upload");
    expect(second.n).toBe(1);
    expect(second.fp).toBe("b27b5fe1");
    expect((second.pair.req.payload as GrpcStreamPayload).n).toBe(1);
    expect(second.pair.req.stream.frames).toHaveLength(2);
  });
});

File: go/conformance_test.go
package xrr_test

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"testing"

	"gopkg.in/yaml.v3"
	xrr "hop.top/xrr"
	execadapter "hop.top/xrr/adapters/exec"
	fsadapter "hop.top/xrr/adapters/fs"
	httpadapter "hop.top/xrr/adapters/http"
	redisadapter "hop.top/xrr/adapters/redis"
	sqladapter "hop.top/xrr/adapters/sql"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

type manifest struct {
	Interactions []struct {
		Adapter     string `yaml:"adapter"`
		Fingerprint string `yaml:"fingerprint"`
		Streamed    bool   `yaml:"streamed"`
		// VerifyFingerprint marks a unary entry whose fingerprint is a
		// computed value: the walker rebuilds the adapter's request from
		// the req payload and recomputes it with the adapter's algorithm.
		VerifyFingerprint bool `yaml:"verify_fingerprint"`
	} `yaml:"interactions"`
}

// TestConformanceFixtures replays spec/fixtures cassettes — proves Go can read
// cassettes produced by any other language port. Entries marked streamed are
// routed through the streaming code path per the manifest extension:
// load+validate, lossless re-emit round-trip, and (for grpc) fingerprint
// recomputation against the filename.
func TestConformanceFixtures(t *testing.T) {
	fixtures := filepath.Join("..", "spec", "fixtures")
	entries, err := os.ReadDir(fixtures)
	require.NoError(t, err)
	require.NotEmpty(t, entries, "no fixture dirs found")

	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		e := e
		t.Run(e.Name(), func(t *testing.T) {
			dir := filepath.Join(fixtures, e.Name())
			manifestPath := filepath.Join(dir, "manifest.yaml")
			data, err := os.ReadFile(manifestPath)
			require.NoError(t, err, "missing manifest.yaml in %s", e.Name())

			var m manifest
			require.NoError(t, yaml.Unmarshal(data, &m))

			c := xrr.NewFileCassette(dir)
			for _, interaction := range m.Interactions {
				if interaction.Streamed {
					conformStreamedPair(t, c, interaction.Adapter, interaction.Fingerprint)
					continue
				}
				conformUnaryPair(t, c, interaction.Adapter, interaction.Fingerprint, interaction.VerifyFingerprint)
			}
		})
	}
}

// conformUnaryPair loads one unary pair through the v1 code path and, when
// the manifest pins the fingerprint, recomputes it from the loaded request
// with the adapter's own algorithm. Loading alone cannot expose a
// canonical-JSON escaping fork — the files load in every port; the derived
// key is what differs.
func conformUnaryPair(t *testing.T, c *xrr.FileCassette, adapter, fingerprint string, verify bool) {
	t.Helper()
	var reqPayload, respPayload map[string]any
	_, err := c.Load(adapter, fingerprint, &reqPayload, &respPayload)
	if !assert.NoError(t, err, "cassette miss: adapter=%s fp=%s", adapter, fingerprint) || !verify {
		return
	}
	got, err := recomputeUnaryFingerprint(c, adapter, fingerprint)
	require.NoError(t, err)
	assert.Equal(t, fingerprint, got,
		"unary fingerprint recomputation must match manifest: adapter=%s", adapter)
}

// recomputeUnaryFingerprint decodes the req payload into the adapter's typed
// request and runs the adapter's Fingerprint over it.
func recomputeUnaryFingerprint(c *xrr.FileCassette, adapter, fingerprint string) (string, error) {
	var resp map[string]any
	switch adapter {
	case "exec":
		var req execadapter.Request
		if _, err := c.Load(adapter, fingerprint, &req, &resp); err != nil {
			return "", err
		}
		return execadapter.NewAdapter().Fingerprint(&req)
	case "http":
		var req httpadapter.Request
		if _, err := c.Load(adapter, fingerprint, &req, &resp); err != nil {
			return "", err
		}
		return httpadapter.NewAdapter().Fingerprint(&req)
	case "sql":
		var req sqladapter.Request
		if _, err := c.Load(adapter, fingerprint, &req, &resp); err != nil {
			return "", err
		}
		return sqladapter.NewAdapter().Fingerprint(&req)
	case "fs":
		var req fsadapter.Request
		if _, err := c.Load(adapter, fingerprint, &req, &resp); err != nil {
			return "", err
		}
		return fsadapter.NewAdapter().Fingerprint(&req)
	case "redis":
		var req redisadapter.Request
		if _, err := c.Load(adapter, fingerprint, &req, &resp); err != nil {
			return "", err
		}
		return redisadapter.NewAdapter().Fingerprint(&req)
	}
	return "", fmt.Errorf("verify_fingerprint: no unary fingerprint model for adapter %q", adapter)
}

// conformStreamedPair exercises the format-layer obligations on one streamed
// pair: parse into the stream model, re-emit losslessly (encoding choice
// free; equality over decoded bytes), and recompute the fingerprint from
// open-time inputs for adapters whose algorithm the spec defines (grpc).
func conformStreamedPair(t *testing.T, c *xrr.FileCassette, adapter, fingerprint string) {
	t.Helper()
	pair, err := c.LoadStream(adapter, fingerprint)
	require.NoError(t, err, "streamed load: adapter=%s fp=%s", adapter, fingerprint)

	out := xrr.NewFileCassette(t.TempDir())
	require.NoError(t, out.SaveStream(adapter, fingerprint, pair))
	again, err := out.LoadStream(adapter, fingerprint)
	require.NoError(t, err, "re-emitted pair must load: fp=%s", fingerprint)
	assertStreamPairEqual(t, pair, again)

	if adapter == "grpc" {
		assert.Equal(t, fingerprint, recomputeGRPCStreamFingerprint(t, pair),
			"fingerprint recomputation must match filename")
	}
}

// recomputeGRPCStreamFingerprint recomputes a grpc streaming fingerprint from
// the pair's open-time inputs: the single send frame for server streams, the
// recorded occurrence ordinal for client/bidi. Reading the informational
// payload n is a conformance-only device — replay recomputes its own counter.
//
// This derives n from each pair's own payload rather than from a counter
// shared across the dir, so the loop above needs no ordering: the manifest's
// order cannot affect any result. That satisfies the spec's ordering rule
// (cassette-format-streaming.md, Manifest Extension) vacuously — there is no
// shared counter for a wrong order to desynchronise.
// TestConformanceManifestOrderIrrelevant pins that property so it cannot
// regress into an order dependence silently.
func recomputeGRPCStreamFingerprint(t *testing.T, pair *xrr.StreamPair) string {
	t.Helper()
	service, _ := pair.ReqPayload["service"].(string)
	method, _ := pair.ReqPayload["method"].(string)
	var msg []byte
	n := 0
	if pair.Req.Type == xrr.StreamServer {
		require.Len(t, pair.Req.Frames, 1, "server stream records exactly one send frame")
		msg = pair.Req.Frames[0].Message
	} else {
		recorded, ok := pair.ReqPayload["n"].(int)
		require.True(t, ok, "client/bidi req payload must record n")
		n = recorded
	}
	open := grpcStreamOpen(pair.Req.Type, service, method, msg)
	fp, err := xrr.StreamFingerprint(open, n)
	require.NoError(t, err)
	return fp
}

// TestConformanceMalformedB64Rejected — the grpc-stream-malformed-b64 pair is
// deliberately absent from its manifest (interactions enumerate replayable
// pairs); harnesses target it by path and assert strict loading fails rather
// than silently discarding the bad characters.
func TestConformanceMalformedB64Rejected(t *testing.T) {
	dir := filepath.Join("..", "spec", "fixtures", "grpc-stream-malformed-b64")
	c := xrr.NewFileCassette(dir)
	_, err := c.LoadStream("grpc", "8dbfb222")
	require.Error(t, err)
	assert.NotErrorIs(t, err, xrr.ErrCassetteMiss)
}

// TestConformanceScalarHazards — sse-text-scalars message_text payloads must
// decode as exactly those characters regardless of YAML scalar resolution.
func TestConformanceScalarHazards(t *testing.T) {
	dir := filepath.Join("..", "spec", "fixtures", "sse-text-scalars")
	pair, err := xrr.NewFileCassette(dir).LoadStream("sse", "66ecc77a")
	require.NoError(t, err)

	want := []string{"on", "12:30", "null", " leading", "trailing ", "  padded  "}
	require.Len(t, pair.Resp.Frames, len(want))
	for i, w := range want {
		assert.Equal(t, []byte(w), pair.Resp.Frames[i].Message, "frame %d", i)
	}
}

// TestConformanceClientStreamRepeat — the spec's scripted n=1 obligation: one
// session, two sequential opens of (files.FileService, Upload, client)
// against grpc-client-stream-repeat as the session dir, expecting the n=0
// then n=1 fingerprints and both conversations replayed.
func TestConformanceClientStreamRepeat(t *testing.T) {
	s := fixtureSession(t, "grpc-client-stream-repeat")
	open := grpcStreamOpen(xrr.StreamClient, "files.FileService", "Upload", nil)

	rep1, err := s.OpenStreamReplay(open)
	require.NoError(t, err)
	assert.Equal(t, "2bebfd6f", rep1.Fingerprint())
	require.NoError(t, rep1.Send([]byte("alpha\n")))
	require.NoError(t, rep1.HalfClose())
	msg, err := rep1.Recv()
	require.NoError(t, err)
	assert.Equal(t, []byte(`{"received_bytes":6}`), msg)
	_, err = rep1.Recv()
	assert.ErrorIs(t, err, io.EOF)

	rep2, err := s.OpenStreamReplay(open)
	require.NoError(t, err)
	assert.Equal(t, "b27b5fe1", rep2.Fingerprint())
	require.NoError(t, rep2.Send([]byte("beta-1\n")))
	require.NoError(t, rep2.Send([]byte("beta-2\n")))
	require.NoError(t, rep2.HalfClose())
	msg, err = rep2.Recv()
	require.NoError(t, err)
	assert.Equal(t, []byte(`{"received_bytes":14}`), msg)
	_, err = rep2.Recv()
	assert.ErrorIs(t, err, io.EOF)
}

// TestConformanceManifestOrderIrrelevant pins the property that makes the
// manifest ordering rule (cassette-format-streaming.md, Manifest Extension)
// hold here: `interactions` is an unordered set, so reversing a manifest's
// entries must not change any conformance result. Go recomputes each pair's
// occurrence n from that pair's own payload rather than from a counter shared
// across the dir, so it is order-independent by construction — this test keeps
// that construction from silently regressing into a shared-counter loop, where
// a wrong order would assign entries each other's n and miss.
func TestConformanceManifestOrderIrrelevant(t *testing.T) {
	fixtures := filepath.Join("..", "spec", "fixtures")
	entries, err := os.ReadDir(fixtures)
	require.NoError(t, err)

	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		e := e
		t.Run(e.Name(), func(t *testing.T) {
			dir := filepath.Join(fixtures, e.Name())
			data, err := os.ReadFile(filepath.Join(dir, "manifest.yaml"))
			require.NoError(t, err)

			var m manifest
			require.NoError(t, yaml.Unmarshal(data, &m))

			// Reverse the manifest: a legal edit under the ordering rule.
			reversed := m.Interactions
			for i, j := 0, len(reversed)-1; i < j; i, j = i+1, j-1 {
				reversed[i], reversed[j] = reversed[j], reversed[i]
			}

			c := xrr.NewFileCassette(dir)
			for _, interaction := range reversed {
				if interaction.Streamed {
					conformStreamedPair(t, c, interaction.Adapter, interaction.Fingerprint)
					continue
				}
				conformUnaryPair(t, c, interaction.Adapter, interaction.Fingerprint, interaction.VerifyFingerprint)
			}
		})
	}
}
